### PR TITLE
Generalise statemachine tests to use the ModelIO class interface

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

**Comment by jorisdral** [comment]
Note that this is still a WIP for now, because I haven't reviewed the diff and cleaned it up

**Comment by jorisdral** [comment]
This should now be ready for review

File: test/Database/LSMTree/Model/Normal/Session.hs [hunk 1]
```diff
@@ -152,13 +152,16 @@ type TableHandleID = Int
 
 data TableHandle k v blob = TableHandle {
     tableHandleID :: TableHandleID
-  , config        :: SUT.TableConfig
+  , config        :: TableConfig
   }
   deriving Show
 
+data TableConfig = TableConfig
+  deriving Show
```
Comment (jorisdral) [inline test/Database/LSMTree/Model/Normal/Session.hs:160]: INFO: We don't use the contents of `SUT.Tableconfig` anywhere, so we can define a simple modelled type

File: test/Test/Database/LSMTree/ModelIO/Class.hs [hunk 2]
```diff
@@ -24,6 +24,8 @@ type IsSession :: ((Type -> Type) -> Type) -> Constraint
 class IsSession s where
     newSession :: IOLike m => m (s m)
 
+    closeSession :: IOLike m => s m -> m ()
+
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/ModelIO/Class.hs:28]: INFO: added this so we can properly clean up sessions after running statemachine tests

File: test/Test/Util/TypeFamilyWrappers.hs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+-- | Type family wrappers are useful for a variety of reasons:
+--
+-- * Type families can not be partially applied, but type wrappers can.
+--
+-- * Type family synonyms can not appear in a class head, but type wrappers can.
+--
+-- * Wrappers can be used to direct type family reduction. For an example, see
+--   the uses of 'WrapTableHandle' and co in the definition of 'RealizeIOSim',
+--   which can be found in "Test.Util.Orphans".
```
Comment (jorisdral) [inline test/Test/Util/TypeFamilyWrappers.hs:14]: INFO: This is the main reason we need these wrappers

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 4]
```diff
@@ -59,56 +74,133 @@ import qualified Test.QuickCheck.StateModel.Lockstep.Defaults as Lockstep.Defaul
...
 import           Test.Tasty (TestTree, testGroup)
 import           Test.Tasty.HUnit (testCase)
-import           Test.Util.Orphans (Wrap (..))
+import           Test.Tasty.QuickCheck (testProperty)
+import           Test.Util.Orphans ()
+import           Test.Util.TypeFamilyWrappers (WrapBlob (..), WrapBlobRef (..),
+                     WrapSession (..), WrapTableHandle (..))
 
 {-------------------------------------------------------------------------------
   Test tree
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:83]: INFO: there is quite a bit of duplication in this section between the `propLockstep*` functions, but I didn't want to define a general property yet (something for later, when these tests become more concrete)

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 5]
```diff
@@ -59,56 +74,133 @@ import qualified Test.QuickCheck.StateModel.Lockstep.Defaults as Lockstep.Defaul
...
+      , Impl.Real.tcBitPrecision = ()
+      }
+
 {-------------------------------------------------------------------------------
   Model state
 -------------------------------------------------------------------------------}
 
-data ModelState = ModelState Model.Model Stats
+type ModelState :: ((Type -> Type) -> Type -> Type -> Type -> Type) -> Type
+data ModelState h = ModelState Model.Model Stats
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:190]: INFO: parameterising the model state by the table handle implementation `h` ensures that we can use the same statemachine for different table handle implementations. It does mean we've had to add a lot of `h` parameters everywhere.

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 6]
```diff
@@ -295,39 +409,56 @@ instance InLockstep ModelState where
...
+  Real monad
+-------------------------------------------------------------------------------}
+
+-- | Uses 'WrapSession' such that we can define class instances that mention
+-- 'RealMonad' in the class head. This is necessary because class heads can not
+-- mention type families directly.
+--
+-- Also carries an exception handle that is specific to the table implementation
+-- identified by the table handle @h@.
+type RealMonad h m = ReaderT (WrapSession h m, Handler m (Maybe Model.Err)) m
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:445]: INFO: each implementation (identified by @h@) could throw different errors, so a handler has to be passed in up front. The handler could probably also be passed into the statemachine framework using a type class

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 7]
```diff
@@ -410,146 +556,162 @@ runModel lookUp = \case
...
+    lookUp' = lookUpGVar (Proxy @(RealMonad h (IOSim s))) lookUp
+
+catchErr ::
+     forall m a. MonadCatch m
+  => Handler m (Maybe Model.Err) -> m a -> m (Either Model.Err a)
+catchErr (Handler f) action = catch (Right <$> action) f'
   where
-    handler :: SomeException -> m (Either Model.Err b)
-    handler (SomeException e) = maybe (throwIO e) (pure . Left) Nothing
+    f' e = maybe (throwIO e) (pure . Left) =<< f e
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:661]: INFO: we use the custom error handler here

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 8]
```diff
@@ -563,26 +725,18 @@ arbitraryActionWithVars _ findVars _st = QC.oneof $ concat [
...
+      -> Var h a
     fromRight = mapGVar (\op -> OpFromRight `OpComp` op)
 
-    -- TODO: improve, more types of configs
-    genTableConfig :: Gen SUT.TableConfig
-    genTableConfig = pure $ SUT.TableConfig {
-          SUT.tcMaxBufferMemory = 2 * 1024 * 1024
-        , SUT.tcMaxBloomFilterMemory = 2 * 1024 * 1024
-        , SUT.tcBitPrecision = ()
-        }
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:584]: INFO: this is replaced by an `Arbitrary (SUT.Class.TableConfig h)` constraint

**Review by jorisdral — COMMENTED** [review]
Some info for reviewers

**Comment by dcoutts** [comment]
Rebased on #34 to fix CI.

**Pull request merged**